third_party/amd/lib/TritonAMDGPUToLLVM/SchedInstructions.cpp

```diff
@@ -181,14 +199,108 @@ struct InsertInstructionSchedHints
     MLIRContext *ctx = &getContext();
     ModuleOp mod = getOperation();
 
-    mod->walk([ctx](triton::DotOp dot) {
-      if (dyn_cast<mlir::scf::ForOp>(dot->getParentOp())) {
+    mod->walk([this, ctx](triton::DotOp dot) {
+      if (auto forOp = dyn_cast<scf::ForOp>(dot->getParentOp())) {
         mlir::OpBuilder rewriter(ctx);
         rewriter.setInsertionPointAfter(dot);
         rewriter.create<triton::amdgpu::InstructionSchedHint>(dot->getLoc());
+        anotateDotOperands(forOp);
       }
     });
   }
+
+  template <typename OpType>
+  llvm::SmallVector<OpType> getUsersOfType(Value value) const {
+    llvm::SmallVector<OpType> concreteUsers;
+    for (auto user : value.getUsers()) {
+      if (auto concreteUser = llvm::dyn_cast<OpType>(user))
+        concreteUsers.push_back(concreteUser);
+    }
+    return concreteUsers;
+  }
+
+  Value rewindConvertLayoutOps(Value value) const {
```

Comment:
Need documentation for this to describe what's it's doing like go through the single use chain of convert_layout op to get the final value after all conversions.

Reply:
Done